Check specific validation error in miner tests

Sjors opened the pull request:
## Problem

`BOOST_CHECK_THROW` merely checks that some `std::runtime_error` is
thrown, but not which one.

Here's an example of how this can cause a test to pass when a developer
introduces a consensus bug. The test for the sigops limit assumes
that `CreateNewBlock` fails with `bad-blk-sigops`. However it can
also fail with bad-txns-vout-negative, if a naive developer lowers
`BLOCKSUBSIDY` to `1*COIN`.

## Solution

`BOOST_CHECK_EXCEPTION` allows an additional predicate function. This
commit uses this for all exceptions that are checked for in
`miner_tets.cpp`:
* `bad-blk-sigops`
* `bad-cb-multiple`
* `bad-txns-inputs-missingorspent`
* `block-validation-failed`

If the function throws a different error, the test will fail. Although the message produced by Boost is a bit [confusing](http://boost.2283326.n4.nabble.com/Test-BOOST-CHECK-EXCEPTION-error-message-still-vague-tt4683257.html#a4683554), it does show which error was actually thrown. Here's what the above `1*COIN` bug would result in:

<img width="1134" alt="schermafbeelding 2017-09-02 om 23 42 29" src="https://user-images.githubusercontent.com/10217/29998976-815cabce-9038-11e7-9c46-f5f6cfb0ca7d.png">
 
## Other considerations

A more elegant solution in my opinion would be to subclass `std::runtime_error` for each `INVALID_TRANSACTION` type, but this would involve touching consensus code.

I put the predicates in `test_bitcoin.h` because I assume they can be reused in other test files. However [serialize_tests.cpp](https://github.com/bitcoin/bitcoin/blob/v0.15.0rc3/src/test/serialize_tests.cpp#L245) also uses `BOOST_CHECK_EXCEPTION` and it defines the predicate in the test file itself.

Instead of four `IsRejectInvalidReasonX(std::runtime_error const& e)` functions, I'd prefer something reusable like `bool IsRejectInvalidReason(String reason)(std::runtime_error const& e)`, which would be used like `BOOST_CHECK_EXCEPTION(functionThatThrows(), std::runtime_error, IsRejectInvalidReason("bad-blk-sigops")`. I couldn't figure out how to do that in C++.

luke-jr commented:
Most of the rejection reasons are not standardised, and we only care that it fails. It might make the most sense to check that it's a `consensus_rule_violation` exception or such, but I'm not sure checking the specific string is useful.

Your example is invalid, because for the `bad-blk-sigops` test, we would not care what the outcome is if the subsidy is broken. Instead, the subsidy should be tested separately, not implicitly.

sdaftuar commented:
Concept ACK.  I think in general checking for the specific error we expect makes it less likely that we inadvertently introduce a bug in the tests; certainly this kind of thing has cropped up in the functional test suite, where bugs introduced into a test harness cause the test to not actually be testing anything, or to be testing the wrong thing.

IMO the downside (in general) to additional specificity is that it adds overhead when making changes to the consensus code rejection reasons, because now all the tests that were hardcoded for a particular reason need updating.  This patch seems pretty small though, and I don't think that should be much of a concern here.  In fact I think we largely already do similar reject-reason-pattern-matching in the functional tests as well.

Sjors commented:
I'll see what I can do about Travis failing on everything but OSX. It's throwing `multiple definition of IsRejectInvalidReason` in a bunch of places.

I agree the example isn't great. I did actually run into it while trying an extremely high value for MAX_BLOCK_SIGOPS_COST (250x) in an [experiment](https://github.com/Sjors/bitcoin/pull/1). In attempt to make the test pass, I increased the transaction generation loops from 1001 to 250001, which obviously(?) took forever to run. I then tried fewer loops to get a sense of performance. To my surprise the test passed for some seemingly random intermediate value. Once I found the actual error message I realized it was an error in the test; it just ran of coins somewhere in the loop.

Sjors commented:
I pushed a few improvements, apparently to the satisfaction of Travis:

1. moved changes from `test_bitcoin.h` to `testutil.h`, which seems more appropriate
2. instead of one function for each validation error type, there's now a `class CheckRejectInvalid` whose constructor takes a string
3. `#define` for validation error strings so they're all in one place. That should make later refactoring easier and also prevent typos.

So now each test looks like this:

```cpp
BOOST_CHECK_EXCEPTION(
    functionThatThrows(),
    std::runtime_error,
    CheckRejectInvalid(REJECT_INVALID_BAD_CB_MULTIPLE)
);
```

Unfortunately as a result of this change, when a test throws a different exception, it no longer shows what that exception was. However, this is easy for a developer to debug: just put `BOOST_CHECK(functionThatThrows())` above the failing `BOOST_CHECK_EXCEPTION ` and run the test again.

Sjors commented:
@sipa and @morcos worked on `miner_tests.cpp` earlier this year.

@morcos since you refactored `CreateNewBlock` in late 2015 (#6898), could you check that I picked the right specific validation errors? Just to make sure the tests weren't already misbehaving.

MarcoFalke commented on src/test/testutil.h:
```diff
@@ -9,7 +9,27 @@
+#define REJECT_INVALID_BLOCK_VALIDATION_FAILED "block-validation-failed"
+
+// BOOST_CHECK_THROW predicates to check the specific validation error
+class CheckRejectInvalid {
```
Maybe move all of this to the cpp?

Sjors commented:
@MarcoFalke  since #11234 just snatched `testutil.h` out from under me, I moved everything into `miner_test.cpp`.

There are a lot of other tests that use `BOOST_CHECK_THROW` without checking the precise error, so it might make sense to revert #11234 at some point.

MarcoFalke commented:
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

Sjors commented:
Done, as well as rebased to latest master.

Sjors commented:
Done, as well as rebased to latest master.

Sjors commented:
I pushed again to fix commit message and code comment (it was saying "BOOST_CHECK_THROW" instead of "BOOST_CHECK_EXCEPTION" in two places).

Sjors commented:
The `p2p-segwit.py` test failed in one environment. Random time-out? All machines passed on my previous force push.
![schermafbeelding 2017-10-03 om 11 16 49](https://user-images.githubusercontent.com/10217/31120475-713a535c-a82c-11e7-969b-cc0c8a6fc21f.png)

MeshCollider commented:
@Sjors probably just an unrelated failure, Travis does that sometimes. I've restarted that test for you.

promag commented on src/test/miner_tests.cpp:
```diff
@@ -26,6 +27,25 @@
 BOOST_FIXTURE_TEST_SUITE(miner_tests, TestingSetup)
 
+// Validation errors used in tests (prevent typos and facilitate refactoring):
+#define REJECT_INVALID_BAD_BLK_SIGOPS "bad-blk-sigops"
```
IMO inline these strings.

Sjors replied:
I used `#define` to make it easier to find all occurrences in the codebase if these rejection reasons ever need to be changed. It also helps against typos, although typos would be immediately obvious when you run the test.

Happy to change it to inline if nobody objects.

promag replied:
The test will fail if there is a typo. Having inline is good because a rename forces you to update the relevant tests and double check everything related.

Sjors replied:
Using inline now.

promag commented on src/test/miner_tests.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "uint256.h"
 #include "util.h"
 #include "utilstrencodings.h"
+#include <boost/algorithm/string.hpp>
```
Move down near other boost includes (try to keep it sorted too).

Sjors replied:
I got rid of this dependency.

promag commented on src/test/miner_tests.cpp:
```diff
@@ -26,6 +27,25 @@
+// BOOST_CHECK_EXCEPTION predicates to check the specific validation error
+class CheckRejectInvalid {
+public:
+    CheckRejectInvalid(std::string const& arg) {
```
```cpp
CheckRejectInvalid(const std::string& reason) : m_reason(reason) {}
```

Sjors replied:
Thanks. If it wasn't obvious, I'm not terribly familiar with C++. :-)

Sjors replied:
Done, and renamed the private variable from `reason` to `m_reason`

promag commented on src/test/miner_tests.cpp:
```diff
@@ -26,6 +27,25 @@
+        reason = arg;
+    };
+    bool operator() (std::runtime_error const& e) const {
+        return boost::contains(e.what(), reason);
```
Use `std::string::find`?

Sjors replied:
That should indeed help with #8670.

Sjors replied:
Done

promag commented on src/test/miner_tests.cpp:
```diff
@@ -26,6 +27,25 @@
+    CheckRejectInvalid(std::string const& arg) {
+        reason = arg;
+    };
+    bool operator() (std::runtime_error const& e) const {
```
Nit, `const std::runtime_error& e`.

Sjors replied:
Done

promag commented on src/test/miner_tests.cpp:
```diff
@@ -264,7 +284,9 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
-    BOOST_CHECK_THROW(AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey), std::runtime_error);
+
+    // This should throw bad-blk-sigops
+    BOOST_CHECK_EXCEPTION(AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey), std::runtime_error, CheckRejectInvalid(REJECT_INVALID_BAD_BLK_SIGOPS));
```
IMO this sounds better:
```cpp
..., std::runtime_error, HasReason("bad-blk-sigops"));
```

Sjors replied:
I can rename `CheckRejectInvalid` to `HasReason`; seems more readable indeed.

promag reviewed (commented):
Agree with better testing the exception.

MarcoFalke commented on src/test/miner_tests.cpp:
```diff
@@ -26,6 +27,25 @@
+        return boost::contains(e.what(), reason);
+    };
+private:
+    std::string reason;
```
µnit: Can be `const`

Sjors replied:
Done (Github needs a check-mark emoticon)

Sjors commented:
I believe I addressed all code style issued raised. Rebased to latest master. I also signed the commit this time.

Sjors commented:
@JeremyRubin it would be great to get your feedback on this PR, given that you're working on these validation errors. I was able to cherry-pick my commit on top of your #11523 without a problem.

JeremyRubin commented:
Concept Ack!

I would think one improvement might be if the strings were not to be literals (as Luke points out, they aren't standardized). I think it's ok for now though.

Sjors commented:
@JeremyRubin the downside of squashing commits is you can't see history :-) I actually used `#define REJECT_INVALID_BAD_CB_MULTIPLE "invalid_bad_cb_multiple"` in an earlier version, but switched back to literal strings  based on @promag's [feedback](https://github.com/bitcoin/bitcoin/pull/11220#discussion_r142387624).

sipa commented:
Concept ACK

MarcoFalke commented:
utACK 59d0ede514f7754ad4df11ffe11f27e2ea63bf3e

Sjors committed:
[Tests] check specific validation error in miner tests

BOOST_CHECK_THROW merely checks that some std::runtime_error is
thrown, but not which one.

One example of how this could lead to a test passing when a developer
introduces a consensus bug: the test for the sigops limit assumes
that CreateNewBlock fails with bad-blk-sigops. However it can
also fail with bad-txns-vout-negative, e.g. if a naive developer lowers
BLOCKSUBSIDY to 1*COIN in the test.

BOOST_CHECK_EXCEPTION allows an additional predicate function. This
commit uses this for all exceptions that are checked for in
miner_tets.cpp:
* bad-blk-sigops
* bad-cb-multiple
* bad-txns-inputs-missingorspent
* block-validation-failed

An instance of the CheckRejectInvalid class (for a given validation string)
is passed to BOOST_CHECK_EXCEPTION.

Sjors commented:
Rebased to account for #11389.
